File: app/src/main/java/com/epam/valkaryne/openmovieapp/data/repository/OpenMovieRepositoryImpl.java
package com.epam.valkaryne.openmovieapp.data.repository;

import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;
import androidx.paging.rxjava2.PagingRx;

import com.epam.valkaryne.openmovieapp.core.model.QueryModel;
import com.epam.valkaryne.openmovieapp.data.api.OpenMovieDataSource;
import com.epam.valkaryne.openmovieapp.data.api.model.MovieInfo;
import com.epam.valkaryne.openmovieapp.data.api.retrofit.OpenMovieApiService;
import com.epam.valkaryne.openmovieapp.domain.OpenMovieRepository;

import io.reactivex.Observable;

public class OpenMovieRepositoryImpl implements OpenMovieRepository {

    private final int NETWORK_PAGE_SIZE = 10;
    private final int PREFETCH_DISTANCE = 5;

    private OpenMovieApiService service;

    public OpenMovieRepositoryImpl(OpenMovieApiService service) {
        this.service = service;
    }

    @Override
    public Observable<PagingData<MovieInfo>> getSearchResult(final QueryModel query) {
        return PagingRx.getObservable(
                new Pager(
                        new PagingConfig(NETWORK_PAGE_SIZE, PREFETCH_DISTANCE, false),
                        () -> new OpenMovieDataSource(service, query)
                )
        );
    }
}

File: app/src/main/java/com/epam/valkaryne/openmovieapp/data/repository/OpenMovieRepositoryImpl.kt
package com.epam.valkaryne.openmovieapp.data.repository

import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import com.epam.valkaryne.openmovieapp.core.model.QueryModel
import com.epam.valkaryne.openmovieapp.data.api.OpenMovieDataSource
import com.epam.valkaryne.openmovieapp.data.api.model.MovieInfo
import com.epam.valkaryne.openmovieapp.data.api.retrofit.OpenMovieApiService
import com.epam.valkaryne.openmovieapp.domain.OpenMovieRepository
import kotlinx.coroutines.flow.Flow

class OpenMovieRepositoryImpl(private val service: OpenMovieApiService) : OpenMovieRepository {

    override fun getSearchResult(query: QueryModel): Flow<PagingData<MovieInfo>> {
        return Pager(
            config = PagingConfig(
                pageSize = NETWORK_PAGE_SIZE,
                prefetchDistance = PREFETCH_DISTANCE,
                enablePlaceholders = false
            ),
            pagingSourceFactory = { OpenMovieDataSource(service, query) }
        ).flow
    }

    private companion object {
        const val NETWORK_PAGE_SIZE = 10
        const val PREFETCH_DISTANCE = 5
    }
}